EndTime: fix overflow when doing large comparisons

lrusak opened the pull request:
This fixes the issues pointed out by @thexai in #22416 

The idea here is to limit the maximum allowable duration of the timer to be `std::chrono::steady_clock::duration::max()` This is typically `std::chrono::nanoseconds::max()` which is about 292 years on the platform I'm working on.

The underlying problem is this
```
std::chrono::milliseconds::max() > std::chrono::nanoseconds::max()
```
This comparison isn't possible without some type of casting. If we cast milliseconds to nanoseconds we overflow. So we could cast nanoseconds to milliseconds (though we may lose precision), my fix is to simply limit the duration to `std::chrono::nanoseconds::max()` (`std::chrono::steady_clock::duration::max()`)as stated above.

This changes the notion of `infinite` in EndTime to be `std::chrono::nanoseconds::max()` and thus we can simplify some methods.

I added a static method `EndTime<>::Max()` which will return the maximum allowable duration that is able to be set in the templated type.

So methods that used `std::chrono::milliseconds::max()` can now use `EndTime<>::Max()`. Using the former won't harm as the duration will simply be truncated in `Set`.

I also added some tests to help with this behavior.

I also added a log message in case we ever see a truncation happen in the future (maybe it should throw instead?) as I believe I fixed the (2) locations that the overflows happened. I'm not sure if anyone noticed the bug in `CGUIListItemLayout` but this fixes the issue there too.

lrusak committed:
EndTime: call Set in constructor

Signed-off-by: Lukas Rusak <lorusak@gmail.com>

lrusak committed:
EndTime: add m_max to truncate based on steady_clock::duration::max() casted to templated type

This is required because we use the default std::chrono::steady_clock::duration
which typically is std::chrono::nanoseconds. When comparing durations the templated type
may overflow when comparing to std::chrono::nanoseconds.

std::chrono::nanoseconds::max() on my platform works out to ~292 years

Signed-off-by: Lukas Rusak <lorusak@gmail.com>

lrusak committed:
CActiveAESink: use EndTime<>::Max() for m_extSilenceTimeout while streaming

Signed-off-by: Lukas Rusak <lorusak@gmail.com>

lrusak committed:
CGUIListItemLayout: use EndTime<>::Max() for m_infoUpdateMillis

Signed-off-by: Lukas Rusak <lorusak@gmail.com>

lrusak committed:
EndTime: remove unused method IsInfinite

Signed-off-by: Lukas Rusak <lorusak@gmail.com>

lrusak committed:
TestEndTime: update test to reflect changes to EndTime

Signed-off-by: Lukas Rusak <lorusak@gmail.com>

lrusak committed:
TestEndTime: add test to test std::chrono::steady_clock::duration

Signed-off-by: Lukas Rusak <lorusak@gmail.com>

fritsch commented:
There is one additional usage of INT_MAX in the code for a timer - I am not at my PC currently, so can't point it out - here care is needed as well.

thexai commented on xbmc/threads/SystemClock.h:
```diff
@@ -32,36 +34,39 @@ template<typename T>
+    if (duration > m_max)
+    {
+      m_totalWaitTime = m_max;
+      CLog::Log(LOGDEBUG, "duration ({}) greater than max ({}) - duration will be truncated!",
```
Better change to LOGWARNING to become always visible

lrusak replied:
updated

thexai commented:
It's actually happening:

`2023-01-13 10:15:44.589 T:21096 warning <general>: duration (128849018820000) greater than max (9223372036854) - duration will be truncated!`

Due:

https://github.com/xbmc/xbmc/blob/e9c08df8289b97e539b8b45540ad42de858fffae/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESettings.cpp#L129

This is more than 4000 years!!!

I recommend change this line to:

`std::numeric_limits<short>::max() `

As 546 hours is enough for audio keep alive "Always"

fritsch commented:
Thanks @thexai - awesome find. Let's make this a WARNING then, then everyone sees it. I would still suggest backporting the proper solution, so that we have a "robust" code-base.

fritsch reviewed (commented):
Please add the WARNING instead of the DEBUG - then it's good to go I'd say.

lrusak committed:
CActiveAESettings: use EndTime<std::chrono::minutes>::Max() for Always silence timeout

This is required because the setting itself is stored as an int. The int represents minutes and when converted
to std::chrono::minutes will overflow std::chrono::nanoseconds which is the units of the timer.

thexai commented:
Seems this is changed to "magic number" that represents ~292 years in minutes. Is also a good solution but then Nexus backport requires all commits in this PR and the previous. Not tested but should work.

fritsch reviewed (approved):
Thanks very much. Should go into Nexus in combination with the minutes-fix, then we have a clean / maintainable base there. Awesome.

fritsch commented:
One feedback from the forum: https://forum.kodi.tv/showthread.php?tid=368027&pid=3131293#pid3131293
